proposals/0021-root-signature-in-clang.md

```diff
@@ -0,0 +1,247 @@
+diagnostics.
+2) Compiler must translate textual root signatures into binary encoded 
+formats.
+3) Compiler must support reading and writing binary encoded root 
+ignatures to and from DXContainer files.
+4) Compiler must validate root signatures before emitting them.
+5) Compiler must have a comprehensive test infrastructure to test all 
+software components.
+
+Non-requirements
+1) Compiler is not required to validate or modify root signatures read 
+from a DXContainer.
+
+Other requirements for the overall effort that need to influence the design:
+1) Enable other frontends to utilize the DirectX backend.
+2) Plan for the evolution of root signatures in future versions.
+
+## Proposed solution
+
+The root signature in a compiler begins with a string in the root signature
+attribute and ends with a serialized root signature in the DXContainer.
+
+To report errors as early as possible, root signature string parsing should
+occur in Sema.
+
+To ensure that a used resource has a binding in the root signature, this
+information should be accessible in the backend to make sure legal dxil is
+generated.
+
+A new AST node HLSLRootSignatureDecl will be added to represent the root
+signature in the AST.
+To avoid parsing root signature string twice, HLSLRootSignatureDecl will 
+save the parsed root siganture instead of the string.
+This will require more work for print and serialization the AST node.
+Since DirectX backend also need to print and serialzation root signature, 
+the code should be able to share between the frontend and backend.
+
+A HLSLRootSignatureAttr will be created when meet RootSignature attribute in
+HLSL.
+Because the RootSignature attribute in hlsl only have the string, it is easier
+to add a StringArgument to save the string first.
+A HLSLRootSignatureDecl will be created and added to the HLSLRootSignatureAttr 
+for diagnostic and saved to the HLSLEntryRootSignatureAttr for clang 
+code generation.
+The HLSLRootSignatureDecl will save StringLiteral instead StringRef for 
+diagnostic.
+
+In clang code generation, the HLSLRootSignatureAttr in AST will be translated
+into metadata to express the layout and things like static sampler, root 
+flags, space and NumDescriptors in LLVM IR.
+
+CGHLSLRuntime will generate metadata to link the metadata as root
+signature for given entry function.
+
+For case compile to a standalone root signature blob, the
+HLSLRootSignatureAttr will be bind to a fake empty entry.
+
+In the case of local or global root signatures, only the HLSLRootSignatureDecl 
+will be created.
+
+Additionally, a new AST node, SubobjectToExportsAssociationDecl, should be 
+added to handle the SubobjectToExportsAssociation statement in HLSL. 
+The SubobjectToExportsAssociationDecl will contain a HLSLRootSignatureDecl and 
+a list of FunctionDecls.
+
+The default root signature version number is 2 which maps to rootsig_1_1.
+With option -force-rootsig-ver rootsig_1_0 could make it to 1.
+
+## Detailed design
```

Comment:
This section is called "Detailed design", but it's currently a list of examples. If this is just examples we should label it as such, but it would also be helpful to have more detailed descriptions of the representations throughout.

Reply:
Changed to Examples.